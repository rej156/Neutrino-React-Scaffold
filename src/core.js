import clone from 'ramda/src/clone';
import uniq from 'ramda/src/uniq';
import isNil from 'ramda/src/isNil';

// Core API
const core = {
  clear: {},
  get: {},
  register: {},
};
export default core;

// Top-level application items
// the initial value determines how items are registered / updated
// - for []: allows any item & register "uniquely" concatenates
// - for {}: only allows object items & register merges
// - for null | undefined: allows any item & register replacea
const coreItems = {
  componentNeeds: {},
  initialState: {},
  persistedStateKeys: [],
  preRenderActions: [],
  reducers: {},
  rootComponent: null,
  router: null,
  routesHandlers: [],
  store: null,
  storeMiddleware: [],
};

const isObject = val =>
  val !== null && !Array.isArray(val) && val instanceof Object && typeof val === 'object';

class CoreItem {
  constructor(key) {
    this.key = key;
    this.initialValue = this.get();
  }

  register = (value, shouldReplace) => {
    const key = this.key;
    const currentValue = this.get();

    const done = updatedValue => {
      coreItems[key] = updatedValue;

      return core.register;
    };

    if (isNil(value)) {
      return done(shouldReplace ? this.initialValue : currentValue);
    }

    if (Array.isArray(this.initialValue)) {
      return done(shouldReplace ? [].concat(value) : uniq(currentValue.concat(value)));
    }

    if (isObject(this.initialValue)) {
      if (!isObject(value) && !shouldReplace) {
        throw new Error(
          `core.register.${key}() expects an object, you passed ${JSON.stringify(value)}.`,
        );
      }

      return done(shouldReplace ? value : { ...currentValue, ...value });
    }

    return done(value);
  };

  clear = () => {
    this.register(null, true);

    return core.clear;
  };

  get = () => clone(coreItems[this.key]);
}

/**
 * All core item keys
 */
export const itemKeys = Object.keys(coreItems);

/**
 * Assign getter & setter functions for core items, e.g:
 * - core.register.store(createStore(...))
 * - core.clear.store()
 * - core.get.store()
 */
itemKeys.forEach(key => {
  const item = new CoreItem(key);

  core.register[key] = item.register;
  core.clear[key] = item.clear;
  core.get[key] = item.get;
});
