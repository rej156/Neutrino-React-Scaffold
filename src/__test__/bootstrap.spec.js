import core, { itemKeys } from '../core'

describe('@monty/bootstrap', () => {
  itemKeys.forEach((registerItem) => {
    core.get[registerItem] = jest.fn(core.get[registerItem])
    core.register[registerItem] = jest.fn(core.register[registerItem])
  })

  const ORIGINAL_NODE_ENV = process.env.NODE_ENV
  const mockCore = core

  const requireBootstrap = (mockFeaturesConfig) => {
    jest.mock('../features.config', () => mockFeaturesConfig)
    jest.mock('../core', () => mockCore)
    return require('../bootstrap')
  }

  beforeEach(() => {
    process.env.NODE_ENV = ORIGINAL_NODE_ENV
    jest.resetModules()
    jest.clearAllMocks()
  })

  describe('.featuresList', () => {
    it('returns list of * (default) features', () => {
      expect(requireBootstrap({
        '*': ['feature-a', 'feature-b', 'featureC']
      }).featuresList).toMatchSnapshot()
    })

    it('returns list of default + NODE_ENV features', () => {
      // default jest NODE_ENV = test
      expect(requireBootstrap({
        '*': ['feature-a', 'feature-b'],
        test: ['foo']
      }).featuresList).toMatchSnapshot()
    })

    it('returns list of development features if NODE_ENV is falsy', () => {
      process.env.NODE_ENV = ''
      expect(requireBootstrap({
        '*': [],
        development: ['dev-feature-1', 'dev-feature-2']
      }).featuresList).toMatchSnapshot()
    })

    it('returns empty list if no features specified', () => {
      expect(requireBootstrap({ '*': [] }).featuresList).toEqual([])
    })
  })

  describe('.bootstrapFeatures(getFeature)', () => {
    let bootstrap
    let featuresList
    const mockGetFeatures = jest.fn()
    const bootstrapFeatures = () => {
      bootstrap.bootstrapFeatures(mockGetFeatures)
    }

    const mockFeatures = {}

    beforeEach(() => {
      bootstrap = requireBootstrap({
        '*': ['featureA', 'featureB', 'featureC']
      })
      featuresList = bootstrap.featuresList

      featuresList.forEach((featureName) => {
        mockFeatures[featureName] = jest.fn(() => ({}))
        mockGetFeatures.mockImplementationOnce(mockFeatures[featureName])
      })
    })

    it('calls getFeature callback with feature name for each feature', () => {
      expect(mockGetFeatures).not.toBeCalled()
      bootstrapFeatures()

      expect(mockGetFeatures).toHaveBeenCalledTimes(featuresList.length)
      featuresList.forEach((featureName) => {
        expect(mockGetFeatures).toBeCalledWith(featureName)
      })
    })

    it('calls getFeature callback once for each feature', () => {
      expect(mockGetFeatures).not.toBeCalled()
      bootstrapFeatures()
      bootstrapFeatures()
      bootstrapFeatures()
      expect(mockGetFeatures).toHaveBeenCalledTimes(featuresList.length)
    })

    describe('feature Root component', () => {
      it('is only registers the first Root once', () => {
        mockFeatures.featureA.mockReturnValueOnce({
          Root: 'RootA'
        })

        mockFeatures.featureB.mockReturnValueOnce({
          Root: 'RootB'
        })

        expect(core.register.rootComponent).not.toBeCalled()
        bootstrapFeatures()
        expect(core.register.rootComponent).toBeCalledWith('RootA')
        expect(core.register.rootComponent).toHaveBeenCalledTimes(1)
      })

      it('is only registered if exported by feature', () => {
        mockFeatures.featureA.mockReturnValueOnce({})
        bootstrapFeatures()
        expect(core.register.rootComponent).not.toBeCalled()
      })
    })

    const testFeatureItems = [
      { itemName: 'getRoutes',
        registerMethodName: 'routesHandlers',
        itemValue: () => {}
      },
      { itemName: 'preRenderActions',
        defaultValue: [],
        itemValue: [() => {}]
      },
      { itemName: 'persistedStateKeys',
        defaultValue: [],
        itemValue: ['stateA', 'stateB']
      },
      { itemName: 'storeMiddleware',
        defaultValue: [],
        itemValue: [() => {}]
      },
      { itemName: 'initialState',
        defaultValue: {},
        itemValue: { stateA: 'A', stateB: 'B' }
      },
      { itemName: 'reducers',
        defaultValue: {},
        itemValue: { reducerA: () => {}, reducer: () => {} }
      },
      { itemName: 'sagas',
        defaultValue: [],
        itemValue: [() => {}]
      }
    ]

    testFeatureItems.forEach(({ itemName, registerMethodName, itemValue, defaultValue }) => {
      describe(`bootstraps feature ${itemName}`, () => {
        const registerMethod = core.register[registerMethodName || itemName]

        it('registers default value if item undefined', () => {
          expect(registerMethod).not.toBeCalled()
          bootstrapFeatures()
          expect(registerMethod).toHaveBeenCalledTimes(featuresList.length)
          expect(registerMethod).toBeCalledWith(defaultValue)
        })

        it('registers exported item for all features', () => {
          const mockFeature = {
            [itemName]: itemValue
          }
          mockFeatures.featureA.mockReturnValueOnce(mockFeature)

          expect(registerMethod).not.toBeCalled()
          bootstrapFeatures()
          expect(registerMethod).toHaveBeenCalledTimes(featuresList.length)
          expect(registerMethod).toBeCalledWith(mockFeature[itemName])
        })
      })
    })
  })
})
