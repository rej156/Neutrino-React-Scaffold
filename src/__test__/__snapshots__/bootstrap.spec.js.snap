// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@monty/bootstrap .featuresList returns list of * (default) features 1`] = `
Array [
  "feature-a",
  "feature-b",
  "featureC",
]
`;

exports[`@monty/bootstrap .featuresList returns list of default + NODE_ENV features 1`] = `
Array [
  "feature-a",
  "feature-b",
  "foo",
]
`;

exports[`@monty/bootstrap .featuresList returns list of development features if NODE_ENV is falsy 1`] = `
Array [
  "dev-feature-1",
  "dev-feature-2",
]
`;

exports[`@monty/bootstrap .featuresList returns list of * (default) features 1`] = `
Array [
  "feature-a",
  "feature-b",
  "featureC",
]
`;

exports[`@monty/bootstrap .featuresList returns list of default + NODE_ENV features 1`] = `
Array [
  "feature-a",
  "feature-b",
  "foo",
]
`;

exports[`@monty/bootstrap .featuresList returns list of development features if NODE_ENV is falsy 1`] = `
Array [
  "dev-feature-1",
  "dev-feature-2",
]
`;
